### PR TITLE
Make application of plotting style optional and only set during plotting

### PR DESCRIPTION
**Pull request by twiecki**
This also changes the plot style a bit.

Looking for feedback, not sure this is the best way to go about this.

### CODE REVIEW

**Comment by gusgordon** [comment]
Looks good to me.

File: pyfolio/plotting.py [hunk 1]
```diff
@@ -30,18 +30,21 @@
...
+    if rc is None:
+        rc = {}
+
+    rc_default = {'lines.linewidth': 1.5,
+                  'axes.facecolor': '0.995',
+                  'figure.facecolor': '0.97'}
+
+    # Add defaults if they do not exist
+    for name, val in rc_default.iteritems():
+        if name not in rc:
```
Comment (ssanderson) [inline pyfolio/plotting.py]: `rc.setdefault(name, val)`

File: pyfolio/plotting.py [hunk 2]
```diff
@@ -30,18 +30,21 @@
...
+def set_context(context='notebook', font_scale=1.5, rc=None):
+    if rc is None:
+        rc = {}
+
+    rc_default = {'lines.linewidth': 1.5,
+                  'axes.facecolor': '0.995',
+                  'figure.facecolor': '0.97'}
+
+    # Add defaults if they do not exist
+    for name, val in rc_default.iteritems():
```
Comment (ssanderson) [inline pyfolio/plotting.py]: This is broken on Py3.

File: pyfolio/plotting.py [hunk 3]
```diff
@@ -30,18 +30,21 @@
 from . import pos
 
 
-def set_plot_defaults():
-    """
-    Sets default plotting/charting colors/styles.
-    """
+def set_context(context='notebook', font_scale=1.5, rc=None):
```
Comment (ssanderson) [inline pyfolio/plotting.py]: IMO `set_context` is a misleading name for something that's still being used as a context manager.  I would just call this `context` and use it as:

```
with plotting.context(...):
    do stuff
```

File: pyfolio/tears.py [hunk 4]
```diff
@@ -64,8 +64,14 @@ def create_returns_tear_sheet(returns, live_start_date=None,
         Daily non-cumulative returns of the first benchmark.
     return_fig : boolean, optional
         If True, returns the figure that was plotted on.
+    set_context : boolean, optional
+        If True, set default plotting style context.
     """
 
+    if set_context:
```
Comment (ssanderson) [inline pyfolio/tears.py]: Is there a reason we're not using a proper `with` block here?  Elsewhere I've used the following idiom for conditional contextmanagers:

```
from contextlib import contextmanager

@contextmanager
def nullctx():
    "No-op contextmanager"
    yield


# Inside some function that maybe wants to use a context manager:
with some_context_manager if some_expression else nullctx():
    # Do context things.
```

File: pyfolio/tears.py [hunk 5]
```diff
@@ -347,8 +384,14 @@ def create_bayesian_tear_sheet(returns, benchmark_rets, live_start_date,
         trading, after its backtest period.
     return_fig : boolean, optional
         If True, returns the figure that was plotted on.
+    set_context : boolean, optional
+        If True, set default plotting style context.
     """
 
+    if set_context:
```
Comment (ssanderson) [inline pyfolio/tears.py]: If this boilerplate is going to be everywhere, we might want to factor it out into a decorator.

File: pyfolio/plotting.py (hunk 1 above)
Comment (twiecki) [inline pyfolio/plotting.py]: neat!

File: pyfolio/tears.py (hunk 4 above)
Comment (richafrank) [inline pyfolio/tears.py]: I've found ExitStack also a good tool for conditional context managers.

Comment (ssanderson) [inline pyfolio/tears.py]: @richafrank I think ExitStack requires contextlib2 as a dependency on Py2?  We may already have that though given that we've got a lot of heavyweight SciPy modules as dependencies.

Comment (richafrank) [inline pyfolio/tears.py]: That's right @ssanderson.  I couldn't say whether it's already required.

**Comment by twiecki** [comment]
@ssanderson I love all of these, will push an update soon.

**Comment by twiecki** [comment]
@ssanderson Incorporated your feedback. I removed the noop context manager and replaced it with an `if` which I think is favorable due do reduced complexity and dependency.

File: pyfolio/plotting.py [hunk 6]
```diff
@@ -30,18 +30,32 @@
 from . import pos
 
 
-def set_plot_defaults():
-    """
-    Sets default plotting/charting colors/styles.
-    """
+def plotting_context(func):
+    """Decorator to set plotting context during function call."""
+    def call_w_context(*args, **kwargs):
```
Comment (ssanderson) [inline pyfolio/plotting.py]: This should be decorated with  `functools.wraps(func)` so we don't lose docstrings and method signatures when inspecting interactively.

File: pyfolio/plotting.py [hunk 7]
```diff
@@ -30,18 +30,32 @@
...
+        set_context = kwargs.pop('set_context', True)
+        if set_context:
+            with context():
+                return func(*args, **kwargs)
+        else:
+            return func(*args, **kwargs)
+    return call_w_context
+
+
+def context(context='notebook', font_scale=1.5, rc=None):
```
Comment (ssanderson) [inline pyfolio/plotting.py]: Is there value in having this accept arguments if it's only called in one place with all defaults?  Do we expect to users to use this standalone somehow?  If so, we should document that expectation.

Comment (ssanderson) [inline pyfolio/plotting.py]: One way to make this useful would be to replace the `set_context` boolean arg below with a `plotting_context` arg, which will be used if set, and will use the defaults here if `None`.  Then users could construct their preferred plotting context with this function and pass it to the plotting functions.

Comment (twiecki) [inline pyfolio/plotting.py]: Yeah, good point. I put that in for future use but there's currently no path to it. Your second idea is interesting but I'm worried that it breaks with the suggested way that these style context managers usually work. If I'm not mistaken, a user couldn't do `with set_style('my_style'): create_tear(set_style=False)` anymore but would have to pass the context in.

Comment (twiecki) [inline pyfolio/plotting.py]: I suppose we could check for `False` but it starts to feel like too many ways to do one thing.

Comment (twiecki) [inline pyfolio/plotting.py]: Actually it is useful already, a user could do:
```python
with pf.plotting.context(font_scale=2):
    pf.create_tear_sheet(set_style=False)
```

I think that's a good way to use it. Will add docs.

Comment (ssanderson) [inline pyfolio/plotting.py]: I think I still have a slight preference for
```
my_context = pf.plotting.context(...)
pf.create_tear_sheet(context=my_context)
```
I don't have a super strong preference either way though.  Thoughts @gusgordon @humdings @justinlent ?

Comment (twiecki) [inline pyfolio/plotting.py]: The other issue I think would be that we would always override a user-set style (e.g. via matplotlibrc file) as you either pass in your own, or use the default one, but you can't disable.

Comment (ssanderson) [inline pyfolio/plotting.py]: Fair enough.  What you've got proposed above makes sense to me then.

**Comment by twiecki** [comment]
Thanks for the review @ssanderson, @gusgordon and @richafrank.

**Pull request merged**